### ISSUE TITLE
Substrate coding conventions I

### ISSUE THREAD

**Issue by bedeho** [Joystream/joystream-landing#36]
# Add your suggestion as a comment!

## Background

Our Substrate code base is starting to get more complicated, and it would be a benefit to harmonise the set of major conventions we follow, so as to follow good best practices, and make reviews more efficient. The goal of this issue is to accumulate suggestions over time, as replies, which we can turn into an eventual convention document. This document can further be turned into rules for our CI linter.

## Major questions that 

- How to decide what is its own module, vs. combining with existing module?
- Modules deserve their own repo? every module?

## Initial suggestions

- All maps must map to `Option` to avoid default construction behaviour of `StorageMap` from allowing us to be lazy about checking ::exists on the same map before lookup. 

- Always try to make a module easily reusable for another runtime, by: 
  - If possible, always provide your own traits for your expectations on other modules, rather than relying on public traits, or traits in the module you are expecting to use.
  - Don't define and implement traits for your own module, see point above. 
  - If possible, strive to implement the business logic of your module separately from the runtime module itself, in a substrate agnostic way.

- Assert as many invariants as possible!

**Comment by bedeho** [comment]
- Avoid containers in `Structs`, e.g. `Vec`, as they will make reading _any_ struct property securely require that the client downloads _all_ struct content, including the potentially large container payload. Instead, move containers to top-level storage, and place in suitable `map` with identifier matching original `Struct`.

**Comment by jfinkhaeuser** [comment]
> Avoid containers in `Structs`, e.g. `Vec`, as they will make reading _any_ struct property securely require that the client downloads _all_ struct content, including the potentially large container payload. Instead, move containers to top-level storage, and place in suitable `map` with identifier matching original `Struct`.

I do agree with this. But the clear downside of this is that you have two data sinks that need to be kept in sync manually. It's IMHO important to point out here that you're buying performance at the cost of introducing more complexity developers need to manage.

**Comment by bedeho** [comment]
I very much agree, and I do not like it. In general, I would be in favor of the opposite actually, to sacrifice performance for reducing the invariant count. 

Why don't we make it that something people should just keep in mind at least, perhaps be more sensitive to it if the container can grow very large? In many cases, it will in practice be very small, in which case its an even worse tradeoff have this rule.

**Comment by bedeho** [comment]
Seriously re-evaluate whether a deletion or moderation operation should be done by doing proper removal from state. Removal from state often will require iteration over possibly unbounded state lengths, and will require introduction of extra state variables to make it practically feasible under even normal circumstances. Failure to deal with this properly leaves the platform open to DoS attacks wich transaction fees are not an effective measure against.

Don't delete, only hide!

**Comment by jfinkhaeuser** [comment]
> Don't delete, only hide!
Luckily, deleting an ID from a `Vec` hides the associated entry quite well.

But this is actually not to do with deletion as such. Your situation is one where one top-level item "contains" a hierarchy of descendants that might all have to be modified. Deletion is actually the simplest modification you can make, you just "unlink" the top-level item from something like the above list of IDs.

Much more complex would be operations where each of the descendants needs to be updated, but kept visible. Or where there is no single, but multiple parent elements.

TL;DR, "don't delete, only hide" is a good start, but I think there's a whole class of issues lurking here.

**Comment by bedeho** [comment]
Any state that applications and end users are too read, must be securely readable, and thus _must_ be part of the state. While an indexing node could on occasion generate such state by tracking events and/or transaction inclusion, this would not yield light client proofs that the end user could verify.
While its tempting to conclude that providing a collection of simple transaction inclusion proofs could be sufficient, it is not, as it does not avoid the problem of censored transactions.

**Comment by bedeho** [comment]
Reply to the following prior idea on _always_ wrapping value type of `StorageMap` in option.

> All maps must map to Option to avoid default construction behaviour of StorageMap from allowing us to be lazy about checking ::exists on the same map before lookup.

While this does prevent the lazyness described, it also now introduces the possibility that one may incorrectly introduce `None` value, which should not even be a possible state. Recall that the original problem was not that the map contained a `None` value, it was that a sloppy developer may neglect to check whether some key actually is in the map.

The cost may not be worth the cure, and it also makes the state less self-describing in order to solve a very basic error.

**Comment by bedeho** [comment]
The state of a module should be constructed in such a way that it is possible for a light client to securely interrogate it in a practical way. This will almost always involve adding some additional state variables which functionally depend on the rest of the state. This does involve introducing an additional runtime invariant (which is evil), but the objective is sufficiently important to warrant it.

**Example:** There is relationship one to many relationships `A->B` for types `A`, `B`. For a light client to know that it had all instances of `B` which corresponded to some instance of `A`, it would have to get all `B` instances to avoid the risk of omission by a full or indexing node. The alternative is for `A` to include a counter field for this relationship. This way, if a light client is given the right number of `B` instance inclusion proofs, it will know that it knows the relationship fully.

**Comment by jfinkhaeuser** [comment]
> > All maps must map to Option to avoid default construction behaviour of StorageMap from allowing us to be lazy about checking ::exists on the same map before lookup.
> 
> While this does prevent the lazyness described, it also now introduces the possibility that one may incorrectly introduce `None` value, which should not even be a possible state.
But that's ok - if a `None` value *is* stored in the map, it is identical to not storing any value. In fact, that makes it possible to use storing a `None` value to effectively delete data (I see the discussion about not deleting, but as discussed there, it's a fairly special case).

IMHO using `Option` solves more problems than it introduces. That's especially true because *writing* state is done via extrinsics, so under the close control of runtime developers, whereas *reading* state is the thing that could go wrong without `Option`.

**Comment by jfinkhaeuser** [comment]
> The state of a module should be constructed in such a way that it is possible for a light client to securely interrogate it in a practical way. This will almost always involve adding some additional state variables which functionally depend on the rest of the state. This does involve introducing an additional runtime invariant (which is evil), but the objective is sufficiently important to warrant it.
> 
> **Example:** There is relationship one to many relationships `A->B` for types `A`, `B`. For a light client to know that it had all instances of `B` which corresponded to some instance of `A`, it would have to get all `B` instances to avoid the risk of omission by a full or indexing node. The alternative is for `A` to include a counter field for this relationship. This way, if a light client is given the right number of `B` instance inclusion proofs, it will know that it knows the relationship fully.

What's wrong with this?
```rust
map A => Vec<B>
// or
map IdOfA => Vec<B>
// or, if B is already in it's own collection even
map IdOfA => Vec<IdOfB>
```

Definitely simpler than manually maintaining extra state. Vectors already contain the counter you're looking for.

**Comment by jfinkhaeuser** [comment]
Somewhat related to avoiding containers in structs, and based on some discussion in https://github.com/Joystream/joystream/pull/45

As a rule, I would avoid creating multiple maps of the same ID to different data, and put the data into a struct. That is, *generally* prefer this:
```rust
struct Foo { A, B }
map Id => Foo // or Option<Foo>, see above
```
to
```rust
map Id => A // Option<A>
map Id => B // Option<B>
```

This should be the rule unless one of the following holds true:
1. There is only an indirect relationship between `A` and `B`, even if the `Id` is the same. An example would be that `ContentId` maps to `DataObject` and `ContentMetadata`, but the first can sanely exist without the second in the state (though only having both makes content discoverable).
1. There is a clear use-case for processing large numbers of `A` without having to refer to `B`. This is an optimization as discussed in the "Don't store containers in structs" case, and *should be an optimization rather than the rule*.

**Comment by bedeho** [comment]
> > The state of a module should be constructed in such a way that it is possible for a light client to securely interrogate it in a practical way. This will almost always involve adding some additional state variables which functionally depend on the rest of the state. This does involve introducing an additional runtime invariant (which is evil), but the objective is sufficiently important to warrant it.
> > **Example:** There is relationship one to many relationships `A->B` for types `A`, `B`. For a light client to know that it had all instances of `B` which corresponded to some instance of `A`, it would have to get all `B` instances to avoid the risk of omission by a full or indexing node. The alternative is for `A` to include a counter field for this relationship. This way, if a light client is given the right number of `B` instance inclusion proofs, it will know that it knows the relationship fully.
> 
> What's wrong with this?
> 
> ```rust
> map A => Vec<B>
> // or
> map IdOfA => Vec<B>
> // or, if B is already in it's own collection even
> map IdOfA => Vec<IdOfB>
> ```
> 
> Definitely simpler than manually maintaining extra state. Vectors already contain the counter you're looking for.

The main problem here is that now I cannot get light client proof of a single entry in the map, I have to download entire map to be sure a single entry exists, that's not viable.

**Comment by jfinkhaeuser** [comment]
> The main problem here is that now I cannot get light client proof of a single entry in the map, I have to download entire map to be sure a single entry exists, that's not viable.
Maybe you have to expand on this problem, because as I read it, this approach actually solves your issue.

**Comment by jfinkhaeuser** [comment]
Ah, I think I understand. You want to know a specific A:B relationship exists?

**Comment by jfinkhaeuser** [comment]
So... if that's the case, then you're right - there's a simple enough solution, though, and that'd be:
```rust
map (IdOfA, IdOfB) => bool
```

I honestly think, though, that this is something of a premature optimization for most cases, and I wouldn't want to raise this to a general rule. If you need to prove something like this, sure, it makes sense.

But consider that every state access from a light client is effectively a WebSocket request, potentially through SSL. That is, you have computational and I/O overhead per *access*. The moment you want to prove that both A:B1 and A:B2 exist (or more precisely at N such proofs), your savings go out of the window because the per-request overhead is higher than the few hundred bytes you need to transfer.

**Comment by bedeho** [comment]
> Ah, I think I understand. You want to know a specific A:B relationship exists?

Securely, yes.

**Comment by bedeho** [comment]
> So... if that's the case, then you're right - there's a simple enough solution, though, and that'd be:
> 
> ```rust
> map (IdOfA, IdOfB) => bool
> ```
> 
> I honestly think, though, that this is something of a premature optimization for most cases, and I wouldn't want to raise this to a general rule. If you need to prove something like this, sure, it makes sense.
> 
> But consider that every state access from a light client is effectively a WebSocket request, potentially through SSL. That is, you have computational and I/O overhead per _access_. The moment you want to prove that both A:B1 and A:B2 exist (or more precisely at N such proofs), your savings go out of the window because the per-request overhead is higher than the few hundred bytes you need to transfer.

We don't need to introduce this sort of state explicilty, the point of storage maps/lists/values ... is that they get encoded in a storage trie in a way that makes generating proofs on membership a direct property of the encoding. The solution that I implicitly had in mind, but perhaps should have written out, is

```
// Types

struct A {
id: AIdType,
number_of_Bs: u32
}

struct B {
id: BIdType
id_of_A: AIdType
}

// Storage
allAs: map AIdType => A
allBs: map BIdType => B
```

Now, if a light client knows an `a:A`, and gets proofs `p_1,...p_{a.number_of_Bs}`, it can be _certain_ that is knows about all Bs in the relationship. Now, obviously, `a.number_of_Bs` is redundant on chain state, you can derive from `allBs`, but for a light client its required, otherwise it cannot detect omission. The only other alternative is to download all of `allBs`.

I hope its clear that this is not an optimization.

**Comment by jfinkhaeuser** [comment]
I do understand your reasoning, though I think you're basing it on light client APIs that I haven't seen yet - maybe I missed them, of course.

The APIs I've dealt with have a couple of hundreds of Bytes of HTTP headers in overhead *for every state query*, and there does not seem to be a query for proofs of state. Which would mean your scheme would require `N * (HTTP Overhead  + sizeof(B))`, while mine might use `HTTP Overhead + (N * sizeof(B))`.

There is no guarantee that my scheme ends up more efficient, mind you. I'm arguing that your scheme is effectively an optimization for when `sizeof(B)` outweighs `N * HTTP Overhead`. Which brings me to my point, it's hard to gauge that your scheme is more efficient without looking at each concrete case, whereas it's easy to gauge that it introduces more state to manage.

All of this goes away, of course, if we find a light client API that lets you batch query proofs only. I just didn't see it (yet).

**Comment by bedeho** [comment]
This has nothing to do about the full node API, it's a conceptual point, and it is not related to query overhead. It has to do with what one can and cannot know securely under different choices of state encoding, all under the assumption that light client proofs are available. We currently do not use them in our testnets, as everyone can trust our hosted full node, so that may explain why you have not seen any proofs in the queries you have looked at.

I think we have reached the limit of what can be resolved in a Github comment thread anyway, so let's resolve out of band :D

**Comment by bedeho** [comment]
All storage fields should be configurable, either explicitly, or built, for the following two reasons

1. In line with the goal of having modules be standalone, a third party developer should conveniently be able to launch a new chain where the initial state of this module is configured in some arbitrary way.

2. Makes it easier to write test cases which cover a broad set of initial states.

**Comment by bedeho** [comment]
We should avoid using `usize`, e.g. as a type field, parameter etc., as the actual size is machine dependant. 

https://doc.rust-lang.org/std/primitive.usize.html

```
The size of this primitive is how many bytes it takes to reference any location in memory. For example, on a 32 bit target, this is 4 bytes and on a 64 bit target, this is 8 bytes.
```

Some rust types require us to deal with it, such as Vec::len, however, we should make an effort to not cast to/from it as much as possible.

**Comment by bedeho** [comment]
We should stick to the "one event per extrinsic" rule, where the purpose of an event is to signal a high-level state transition  (which may be complex), rather than whether a specific part of the state is impacted or not.

## Example

Multiple parts of the state may be changed as part of a single extrinsic, e.g. adding a new thread may involve also introducing a new initial first post, where adding a post normally would be its own separate event associated with an extrinsic explicitly for this purpose. In 

## Rationale

Runtime code will eventually get very complex, and it will be very easy to forget to synch the set of events associated with a given extrinsic as the state gets more complex, and the extrinsic itself becomes more complex. It also will require more blockspace, taking away throughput and history storage space. 

It's better to offload the burden of listening to the right high-level events, and understanding the implications about what part of the state is changed, onto off chain client applications, which then have to make sure to subscribe to the appropriate set of relevant events. So in the example, a client application would have to know that, a new post may be introduced into the forum also as a consequence of a new thread, not only directly adding a post on its own.

**Comment by jfinkhaeuser** [comment]
I don't think the "one event per extrinsic" rule should be a rule, more of a guideline.

In general, I agree with the rationale. But the rationale only works as long as extrinsics do one thing only. That's largely a good idea to follow, too, so most of the time this rationale would probably apply.

The downside is UX. We're already at a point in uploading files where pioneer has to call several extrinsics for an upload, and if it does so in sequence - as the design of the extrinsics sometimes requires - this will take several block generation periods to finish. Since each period takes 6s currently, that'll leave a simple upload doing pretty much nothing for 12+ seconds - terrible UX, in other words.

We've partially circumvented this issue by having an extrinsic do more than one thing, but that's not the best of design... at least not until you divide extrinsics into higher- and lower-level operations. Low-level operations should definitely do one thing only, and generate at most one event.

But there is a strong UX argument for introducing higher-level extrinsics that effectively wrap lower-level ones (in practice, use traits to execute the same code as several lower-level ones rather than invoking the extrinsics directly, but that's sort of an implementation detail).

The example here with uploading files would be to create a `DataObject` for uploading the actual file data to a *Liaison*, and simultaneously adding unpublished/draft metadata to the state referencing that `DataObject`. Saving a 6+s wait period is very much better for the UX, and almost certainly worth having an extrinsic emit two separate events for the two seperate sub-steps it performs.

TL;DR distinguish between high-level and low-level extrinsics; the former are for UX and may emit multiple events, the latter are for module APIs and should emit at most one.

**Comment by bedeho** [comment]
Use `linked_map` which has an iterable key set, rather than `map`. While this map type also allows on chain iteration, we primarily should use it for off-chain iteration.

The only penalty is storing the extra links, which is probably O(`#keys`), so if the use case for the map may have _a lot_ of entries, it could be worth reconsidering, however, lookup times are unaffected.

**Comment by bedeho** [comment]
When writing tests, set up the initial state of interest by explicitly configuring it, not through a sequence of simulate transaction calls.

**Comment by bedeho** [comment]
Library code, tests, mocks and fixtures, should live in separate files.

**Comment by bedeho** [comment]
All modules should have their own folder, or be in a folder that colocates a set of related modules.

**Comment by bedeho** [comment]
Cargo.lock should not be tracked in library repos

https://github.com/rust-lang/cargo/issues/315

**Comment by bedeho** [comment]
My hunch is that for modules without any extrinsics, there should also be no events. Any client module should have its own specific events. This seems cleaner, and also avoids duplication of events which would otherwise fredentuly be the case.

This also opens up another question, what about the possible need for events in response to on_finalized processing. I think the symmetric treatment there is instead to have appropriate callback functions that are called to signal something of interest, and then a user can layer events on top, which again may be more complex. The callback could be a global one configured through the module Trait configuration, or it could be supplied by a user call and stored.

**Comment by bedeho** [comment]
Structs should not have identifier information, keep it in the container (e.g. map). This redundant state can really get you into trouble if it gets out of synch with the container key.

**Comment by bedeho** [comment]
If a module is not plausibly constitute a candidate for third party runtimes, then all trait dependencies configured should be using third party traits that are introduced in scope through a direct dependency on the relevant crate/package.

**Comment by bedeho** [comment]
If you are updating a value that exists, either in a map or a top-level storage variable, then try to use the `mutate` method rather than get/insert|write, it is semantically clearer for the reader. Parity confirms there is no performance benefit.

**Comment by bedeho** [comment]
Granular errors for extrinsic and public methods:

Attempt to use as fine-grained a resolution in the error type returned for any method public method, be it dispatchable or not. 

For strings, this means not collapsing lots of different issues into one literal case, e.g. we prefer
`CannotAddWithLowBalance`, `CannotAddWithoutSignature`, `CannotAddOnOddBlockNumber` over a single `CannotAddNow`.

For type-safe responses, this means covering all possible cases as different genuine states, e.g. by fully loading enums with parameters etc.

This all contributes to the writing of a deeper unit test coverage of error cases, as it is clear that certain execution flows are not being covered.

**Comment by JoshOrndorff** [comment]
> I don't think the "one event per extrinsic" rule

I agree

> But there is a strong UX argument for introducing higher-level extrinsics

In some cases it's more that just UX. In my [tictactoe game](https://github.com/JoshOrndorff/tictactoe) I previously had `take_turn` and `claim_win` as two separate extrinsics, each of which emitted a single event. The [problem](https://github.com/JoshOrndorff/TicTacToe/issues/5) is that taking a final move and claiming a win should happen atomically. If they don't I can make a winning move, but my opponent can take an additional move and claim the win before me by front-running.

The solution in my case was to have a single higher-level extrinsic `take_winning_turn` which does both atomically. It's the same solution described above, but I wanted to point out that it's more than just UX.

**Comment by JoshOrndorff** [comment]
> My hunch is that for modules without any extrinsics, there should also be no events.

I see what you're saying, but I disagree. Checkout my [simple feedback module](https://github.com/JoshOrndorff/Marketplace/blob/master/node/runtime/src/simple_feedback.rs). It has no dispatchable calls (eg extrinsics) but it does have a call that can `rate` that can be made from other runtime modules that compose with it. The reputation system doesn't make sense to use on its own, only in conjunction with other modules. But I still emit an event when a rating is placed.

I'm open to suggestions if you think there's a better design here.

Also hope you don't mind me jumping in the middle of your convo. Just thought there were lots of good ideas floating around here.

**Comment by bedeho** [comment]
Hi @JoshOrndorff , thank you for weighing in, we appreciate it.

> The solution in my case was to have a single higher-level extrinsic

Why does the need for such a single high level extrinsic, which I agree with, imply that you cannot follow the suggested one event rule?  The original post suggests that in this context, there would be a new event, e.g. named `WinningTurnTaken`, and that any observer of this module should understand that semantically this involves both a turn and claiming a win.

I guess I am not seeing the problem? If you take the opposite approach, of now having this extrinsic emit two events, e.g. called `TurnTaken` and `Won`, then you are moving in the direction of the problem described. Now, in this particular module it may end up not being very severe, in particular, if this is the only case, but the problem described becomes increasingly real when the module grows in complexity, and one keeps long term maintainability by a diverse set of participants in mind.

> I'm open to suggestions if you think there's a better design here.

I think its no big problem to emit an event like that, but in practice I think its going to be duplicating effort here, because as you say, this action only makes sense in the context of a broader runtime. So in a particular runtime, there may be all sorts of extra information you may want to combine with such a rating notification, and there is no way of injecting that into the lowest level event you have coupled to your rating system. This means one will end up with two events for the same semantic concept,
which has downsides in terms of resources and clarity to developers working on off-chain applications consuming these events.

What do you think? Do you see these as real downsides, if so, are there counterbalancing upsides?

**Comment by bedeho** [comment]
Due to various artifactual requirements induced by Substrate, we can at times be required to implement the `Default` trait for various traits, even when there is no well defined meaning to default construction.

In those cases, do _not_ use this trait in our code, as it gives the misleading indication that we need this property for our own pruposes. One day we will hopefully drop them. Instead mark the implementation clearly as not being permanent, e.g. here

```Rust
/// OpeningStage must be default constructible because it indirectly is a value in a storage map.
/// ***SHOULD NEVER ACTUALLY GET CALLED, IS REQUIRED TO DUE BAD STORAGE MODEL IN SUBSTRATE***
impl<BlockNumber> Default for ApplicationStage<BlockNumber> {
    fn default() -> Self {
        ApplicationStage::Active
    }
}
```

**Comment by mnaamani** [comment]
> Don't use `Default` trait if you are forced to impl

So I don't disagree with the suggestion and I'm certainly aware of this annoying side affect, but of course I have been forced to implement the Default trait for types that I know have no real meaning to be default constructed, but I failed to make a point about it in the code comments.

So I think your suggestion is appropriate for the value type in storage maps (because we are choosing to avoid wrapping it in an `Option` [just to get around this artifact](https://github.com/Joystream/joystream/issues/36#issuecomment-493971007)) 

I'll just add that when we do inevitably add a Default trait wether explicity as above or by deriving it the same issues spreads to the fields of the type.

The best approach for fields where again the default value isn't meaningful, we should try to make them `Option`s at the same time trying to avoid allowing possible "invalid" states when there is inter-dependence between fields. In the following example `Type` is a value in a map and so must have a default value:

If it doesn't make sense for `a` to be `Some` and `b` to be `None` simultaneously, (an invalid state) in this scenario:
```
struct Type {
     a:  Option<A>,
     b:  Option<B>,
}
```
We want to avoid introducing a default trait for `A` which would happen if we turn it into:

```
struct Type {
     a:  A,
     b:  Option<B>,
}
```

So instead do:

```
struct Type {
     c:  Option<(A, B)>
}
```

**Comment by bedeho** [comment]
# Do not overload return types

Lets say you have a public method in your module, or in the future this would also apply to extrinsics when success results can be any type:

```Rust
pub fn my_public_module_method(..) -> Result<MY_SUCCESS_TYPE, MY_ERROR_TYPE> {
...
let good_outcome = Self::ensure_something(...)?;

}
...
fn ensure_something(...) -> Result<SOME_OTHER_SUCCESS_TYPE, MY_ERROR_TYPE> {
.... 
}
```

Unless it is genuinely the case that `ensure_something` can fully return all errors represented by `MY_ERROR_TYPE`, this should be avoided. This is tempting, as its easy and fast, and its is in effect what we are currently doing for extrinsics since the global error type is always `&'static str`.
However, **this is not safe**, as `ensure_something` can at some point mistakenly return an invalid value.

Instead, if the error domain of `ensure_something` is different, even trivially, then introduce a new explicit representation `ENSURE_SOMETHING_ERROR_TYPE`, and use the `std::convert::From` trait to allow seemless conversion on the fly.

**Its slightly more work, but compile time safety in the runtime is worth almost any level of one time pain.**

**Comment by bedeho** [comment]
# `Option` type over `0`

In some cases, the value `0` can be overloaded to mean the absence of something or some event. For example, if something is supposed to happen in zero blocks, or zero amount of tokens are staked. In such cases, opt for modeling the type as `Option<T>` for clarity.

**Comment by bedeho** [comment]
# Do not overload return types II

A perhaps even cleaner way of doing this is to make `ensure_something` into a macro, rather than a function, this way, one can avoid having to introduce lots of private types as in the first suggestion, and instead directly say

```Rust
ensure_something!(..., ERROR_VALUE1, ...);
```

**Comment by bedeho** [comment]
# Explicit types for variants with parameters

There is the temptation to just have an anonymous inlined parameter list, in particular if you just have a single parameter. So doing

```Rust
enum Foo {
...
MyCase(T)
...
}
```

or often also the inlined anonymous struct version

```Rust
enum Foo {
...
MyCase: {
  field_1 : T_1,
  field_2: T_2, 
  ...
}
...
}
```

This however makes it harder to have clean signatures for pure functions that compute new values for these cases. So for example, you may want to have a function `new_my_case(...)` which gives you a new set of values for the `MyCase` variant value. Now you have to either do

```Rust
new_my_case(...) -> (T_1, T_2, ...)
```

or the truly horrible

```Rust
new_my_case(...) -> Foo
```

where all safety is lost.

**Comment by bedeho** [comment]
# Asserting invariants in production runtime code

Something clever here... tough issue.

**Comment by bedeho** [comment]
# Make new `ensure` abstraction if you need to return a value

If you just want to test a condition, then its fine to just inline condition using native `ensure!`, e.g

```Rust
ensure!(x > y, ...)
```

, so long as the condition is not too complex, and you do not end up repeating yourself too badly.

Otherwise, or if you also need to return a value, then you must ensure a new ensure abstraction, e.g.

```Rust
let outcome = ensure_something!(...)?;
```

**Comment by bedeho** [comment]
# WIP: Side effect free helper/utility routines

NB: I just saw this pattern in my code, the proposal is quite invasive, so approach with caution.

**Comment by bedeho** [comment]
# Avoid matching Christmas trees

When we have state composed over multiple levels, we may easily end up with something like this

```Rust
match .. {
  => {
	match .. {
	  => {
	  	match .. {
		  => {
		  ...
		 _ => {..}
		}

	 _ => {..}
	}
 _ => {..}
}
```

This is hard to read for multiple reasons, one being that you have to make long jumps to see how all cases in each match are being processed. It also prevents making each matching step into its own reusable computation.

An alterantive that remains without side-effects and is much easier to read, as well as make resuable, would be

```
Rust

let x_1 = foo_1(...);

let x_2 = foo_2(..., x_1, ...);

...

let x_N = foo_N(....);
```

where `foo_i` typically will be some sort of _ensure_ statement, if you are in a public method or extrinsic, or it can applications of `Iterator::filter_map`, if you are processing a collection through a sequence of steps, etc. All of them need not actually be methods, they can be inline computations, the main point still stands as long as they do not take the form of the matching christmas tree above.

**Comment by bedeho** [comment]
# `mutate` over `insert` for updates II

There was initially [this](https://github.com/Joystream/joystream/issues/36#issuecomment-534170611) idea to use mutate when you meant mutate

However, this invites writing code with side effects inside the mutator. The alternative is to reconstruct the full object you are updating, and then write it back with an insert statement. 

Less sure what the rule should actually be here.

**Comment by bedeho** [comment]
# Result::Err is for genuine bad inputs, not negative answers

Suppose there was a  `can_do_foo(id: T::PersonId) -> Result<T, E>` to check whether a person, identified with `id` ,could do something. The fact that they could not do it should not be signalled through `E`, but rather `T`. `E` is for genuine errors due to bad inputs, such as `id` not being a valid identifier for any person.

**Comment by bedeho** [comment]
# Assertions never have side-effects

At times, one may be tempted to do this

```Rust

assert!(... some code with side effects, e.g. a function all ... );
```

But this is not great, because the reader typically will tune out contents of assertions, assuming they do nothing that needs to be tracked. Instead, do

```Rust
let val = ... some code with side effects, e.g. a function all ...
assert!(... proposition with val ... );
```

**Comment by mnaamani** [comment]
# Avoid type aliased `u16` as extrinsic parameter type

If at all possible avoid using `u16` in the runtime.
If you must use it as the type for an extrinsic/dispatchable call parameter, do not type alias it. It seems to confuse polkadot-js/api and results in an invalidly signed transaction when decoded by a substrate node.

**Comment by bedeho** [comment]
# can_ checker for extrinsics/methods with Imbalance arguments and Result return value

When a call is made to a method 

```Rust
fn my_method(_: Imbalance) -> Result<_,_>
```

then one cannot do the normal `my_method(my_imbalance)?` in the caller scope because there would have to be some cleanup in the client code to neutralize the imbalance side-effect. One way to resolve this is to introduce a new

```Rust
fn ensure_can_call_my_method(_: Imbalance) -> Result<X,_>
```

which simply checks whether the call to `my_method` work (i.e. return `Ok`). Notice that the return type `X` probably should be introduced, in order to make `ensure_can_call_my_method` reusable by `my_method`, which requires exposing relevant information read & processed from storage.

**Comment by bedeho** [comment]
# Aggressively `derive` standard traits for public types 

If types are needed in some client code that lives in another crate, the Rust coherence rule will make life hard for this downstream developer. So just derive as more than what immediately comes to mind, e.g. `#[derive(Eq, PartialEq, Clone, Debug)]`

**Comment by bedeho** [comment]
# RuntimeDebug over Debug trait

[derive(RuntimeDebug)] also derives Debug for you. But for a wasm build this resolves to a Debug implementation that does not print anything, but it makes the compiler happy

**Comment by bedeho** [comment]
# Parametrise with `T::Trait`

When defining a new type, always parametrize by therruntime module trait `Trait`, _not_ the individual types which will later be selected, for the following two reasons

1.  The signatures which result become extremly long

2. Subtle minor mistakes in instantiation result in error messages which are so long and opaque that it dwarfs any possible benefit.

**Comment by bedeho** [comment]
# Event policy

1. Any code path which mutates state _must_ have an event emitted in it, hence, all extrinsics must have an even triggered.

2. The playload for an event _must_ be the minimal number of parameters such that anyone who knows the state would be able to compute the new state solely from the looking at the extrinsic (and its payload) and the event payload.

**Comment by bedeho** [comment]
# Reusable GenesisConfig builders

As we start writing new modules that rely on existing modules that we have written, the new tests may increasingly require the capability to set up initial state conditions for these dependent modules. To avoid having to redo this multiple times, it's better that each module always assumes their genesis configuration building to be a service for external consumption. This builder should not be in the mock module, as not really mocking, and its not even only for tests. Prior we have only had a builder for the entire test externality, but this builder level is not reusable, it must be the genesis configuration.

**Comment by bedeho** [comment]
Tests must assert events being emitted.

**Comment by bedeho** [comment]
# StorageValues of type Option<_> not configurable

If they are, then then `GenesisConfig` will have `_::default()` value, which will almost always have some unintended semantic. Instead, if tests or genesis block require some explicit value, then it must be set manually.

**Comment by shamil-gadelshin** [comment]
I have a proposition regarding invariant asserts in substrate modules (on hiring example).

We have multiple entries of assert!(...) in our code. In worst case small error deeply rooted in the code will cause repeating crush of the whole platform (and we are not talking about malicious users yet).

Parity warns us about panics:
https://substrate.dev/docs/en/tutorials/tcr/tcr-best-practices

> Never panic - Your runtime module functions should not panic at all. Panics can make the chain prone to attacks. If a runtime function panics during execution, all state changes will be reverted. The user will not be charged anything for this execution. This can be repeated and thus can become an attack vector. The user can do a DoS attack on a panicking node because they won't be charged for any runtime execution that panics. The best approach is to detect situations where a panic may occur later and early-exit in a graceful manner to minimize computation and state inconsistencies. If it's impossible to detect a potential panic without first doing substantial computation, then ensure that the transactor first pays some fee (that can possibly be returned if all goes well) before that computation is done.


Consensys suggests the solution:
https://consensys.github.io/smart-contract-best-practices/software_engineering/#circuit-breakers-pause-contract-functionality

> Circuit Breakers (Pause contract functionality)
> Circuit breakers stop execution if certain conditions are met, and can be useful when new errors are discovered. For example, most actions may be suspended in a contract if a bug is discovered, and the only action now active is a withdrawal. You can either give certain trusted parties the ability to trigger the circuit breaker or else have programmatic rules that automatically trigger the certain breaker when certain conditions are met.


Example from hiring:

```
fn get_opt_stake_amount(stake_id: Option<T::StakeId>) -> BalanceOf<T> {
        stake_id.map_or(<BalanceOf<T> as Zero>::zero(), |stake_id| {
            // INVARIANT: stake MUST exist in the staking module
            assert!(<stake::Stakes<T>>::exists(stake_id));

            let stake = <stake::Stakes<T>>::get(stake_id);

            match stake.staking_status {
                // INVARIANT: stake MUST be in the staked state.
                stake::StakingStatus::Staked(staked_state) => staked_state.staked_amount,
                _ => panic!("stake MUST be in the staked state."),
            }
        })
    }

```
Two invariants in this small function:
- stake should exist
- stake status should be Staked

If we got hierarchy:
- Joystream runtime
- Content Directory Working Group module
- Hiring module
- Opening
- Application
- Stake 

Possible error in Stake logic will crush everything.

What if we limit the possible effect?

I suggest to introduce a concept of broken entity state.

In case of broken stake, stake and application become broken and should be discarded using the ensures.
Other entities will remain active. We would have working current opening and other openings as well as remaining platform.

Worst case scenario for hiring - major invariant breaks - whole module becomes inactive as well as Content Directory Working Group module, but other modules continue to work!

We can check the invariant and in broken case save entity_id and the reason.

```
fn get_opt_stake_amount(stake_id: Option<T::StakeId>) -> Result<..> {
        stake_id.map_or(<BalanceOf<T> as Zero>::zero(), |stake_id| {
            
            ensure_not_broken!(stake_id);

	    // INVARIANT: stake MUST exist in the staking module
            if (!<stake::Stakes<T>>::exists(stake_id)) {
            	let reason = "stake {stake_id} does not exit";
            	set_entity_as_broken_and_return_error!(stake_id, reason);
            }

            let stake = <stake::Stakes<T>>::get(stake_id);

            match stake.staking_status {
                // INVARIANT: stake MUST be in the staked state.
                stake::StakingStatus::Staked(staked_state) => Ok(staked_state.staked_amount),
                _ => {
                	let reason = "stake {stake_id} status is not Staked";
	            	set_entity_as_broken_and_return_error!(stake_id, reason);
                }},
            }
        })
    }

```
Downsides:
- additional code complexity
- module client should be aware of this concept
- what to do with broken entities?

**Comment by shamil-gadelshin** [comment]
Do we really need to ensure_* macros to be exported?

```
#[macro_export]
macro_rules! ensure_opt_unstaking_period_is_ok {...}
```

I didn't find any macro usage of hiring module in Content Director Working Group module.

As well as no stake::ensure_* calls in hiring module regarding to stake module macro exports.

I suggest to use this notation to use macro in lib.rs:

```
#[macro_use]
mod macroes;

```
And delete all unnecessary #[macro_export] directives.

It allows to reduce API surface for the module and improve module incapsulation.

Related question:
https://stackoverflow.com/questions/26731243/how-do-i-use-a-macro-across-module-files

**Comment by bedeho** [comment]
I do not recall the use case for that particular macro, but if its not plausible that it be used in external dependency, such as a test or client code, then yeah just make non-public.
I think this may have been me being sloppy because I did not take the time to clarify the details of macro visibility, I think it was quite nuanced.

Also in general, try to add entries that are suggestions for general DOs and DONTs, this post borders on using the issue just as a general QA thread. I think that there is no point to have a rule about this visibility issue: only things that are useful to the public should be visible. Whether this particular macro is so is not important.

**Comment by shamil-gadelshin** [comment]
I place it in the Coding Convention Suggestions because it affects not only hiring module, but other parts of the system as well. Stake module exports macro as well.
It looks for me as general enough.

**Comment by bedeho** [comment]
# bitflags! for key type of maps

For some reason, Substrate appears to not support encoding enums as map keys. 

Rather than doing some manual fix, try using this macro

https://docs.rs/bitflags/1.2.1/bitflags/

**Comment by bedeho** [comment]
# Do panics & asserts, but only in debug mode

Having asserts and panics in runtime code, but only active in debug mode, has two benefits

1.**Clarity:** It makes it easier for the reader to be aware of the relevant invariants at play.
2.**Loud failure:** it makes it easy to detect bugs when running integration tests or similar complex regression testing.

While having panics is discouraged in production in the official Substrate docs, really only because it makes the system vulnerable to possible DoS attacks, this mixed approach of debug only asserts seems like a perfectly safe and productive convention.

**This idea needs peer-review from Parity devs**

**Comment by bedeho** [comment]
# WIP: Struct fields of dynamic size should be lifted into top level storage maps

_Note: Writing this down as a WIP convention, not totally sure yet._

Take the example where you have

```Rust

struct Person {
..
 payload: Vec<u8>,
}

/// Storage
Persons get(persons): map Id => Perspon;
```

In this case, in particular if `payload` can get quite big over time, and even more so if `payload` is not read frequently in any extrinsics, then instead do

```Rust

struct Person {
..
}

type PersonPayload = Vec<u8>;

/// Storage
Persons get(persons): map Id => Perspon;

PersonPayloadById get(person_payload_by_id) : map Id => PersonPayload
```

This is a few possible upsides

- Easier to make extrinsics safe in terms of DoS
- Reduces possible required fee level for lots of operations
- Makes processing faster under normal circumstances
- Makes runtime migrations possibly much faster, as far less data may need to be read from the database.

The main con is that reduces type safety by introducing a new invariatn tht must be maintained, namely that the key set of `Persons` and `PersonPayloadById` must be in synch.

**Comment by siman** [comment]
> WIP: Struct fields of dynamic size should be lifted into top level storage maps

The next step would be to move payloads to offchain storage :) In this case, all advantages that @bedeho specified for XPayloadById would be even stronger.

**Comment by shamil-gadelshin** [comment]
Sounds reasonable. 

But what if expected dynamic-size field is insignificant to overall struct size? Should we always use this pattern to provide consistent approach to dynamic-size fields or there will be some empiric rules (limits), eg. if possible dynamic-size field size greater than 20 bytes then use pattern (or greater than size of remaining struct)?

**Comment by bedeho** [comment]
> But what if expected dynamic-size field is insignificant when compared to overall struct size? Should we always use this pattern to provide consistent approach to dynamic-size fields or there will be some empirical rules (limits), eg. if possible dynamic-size field size greater than 20 bytes then use pattern (or greater than size of remaining struct)?

This was my initial thought also, however in most cases, the length of the field is controlled by a separate tunable storage parameter, in which case one cannot really have any guarantees at implementation time. I suppose the use case may have some very strong implications, say for example if it's a human-readable user name, that is unlikely to ever be more than 20 bytes perhaps. 
So I am not sure, perhaps we should have a convention which is sensitive to this.

> The next step would be to move payloads to off-chain storage :) In this case, all advantages that @bedeho specified for XPayloadById would be even stronger.

This def. has very strong merit, it's just the practical complexity of making sure that you have robust off-chain infrastructure which will always have it available. Perhaps we can find some sort of general solution for this at a later time, e.g. using this when it is ready

https://github.com/playproject-io/datdot

Alternatively, we could just use the built-in storage system in Joystream, but I do worry about making that complicated in order to cover many disparate uses cases.

**Comment by shamil-gadelshin** [comment]
### How to test on_finalize() in runtime modules

When we test on_finalize() or other methods in runtime modules sometimes we need to fast forward several blocks.

As [Parity suggests](https://substrate.dev/docs/en/next/development/module/tests#block-production) we should create and use `run_to_block()` function:

```
fn run_to_block(n: u64) {
    while System::block_number() < n {
        ExampleModule::on_finalize(System::block_number());
        System::on_finalize(System::block_number());
        System::set_block_number(System::block_number() + 1);
        System::on_initialize(System::block_number());
        ExampleModule::on_initialize(System::block_number());
    }
}
```

on_finalize() invocation can be:

```
fn run_to_block_and_finalize(n: u64) {
    run_to_block(n);
    <Module as OnFinalize<u64>>::on_finalize(n);
}
```

**Comment by shamil-gadelshin** [comment]
### Runtime module Cargo.toml header 

I suggest to create a standard way of composing Cargo.toml header for runtime modules:

```
[package]
name = 'substrate-NAME-module'
version = '1.0.0'
authors = ['Joystream contributors']
edition = "2018"
```

Each module has separate versioning.
Name should have 'substrate' prefix which is an indicator of reusability as substrate framework module.

**Comment by shamil-gadelshin** [comment]
### Monorepo after Rome

We should start gathering runtime modules in monorepo after the Rome release.
Dependencies between crates can be resolved with relative paths. 

[Example from Parity Substrate runtime.](https://github.com/paritytech/substrate/blob/master/bin/node/runtime/Cargo.toml)

**Comment by juniuszhou** [comment]
### Define the all data member of struct as public in runtime module
When I implemented the data migration, I found I can't access some data member defined in forum module because they are private. They are not accessible by other module.

**Comment by juniuszhou** [comment]
### Name the runtime module name with version
When I implemented the data migration, it is better to give the new forum name with version info. Since version not given in old module, I don't how to name the new one. Upgrade may be frequently happened in the future, so it is easier to track the changes.

**Comment by bedeho** [comment]
This has gotten too long, will summarize into a document to be introduced in this PR

https://github.com/Joystream/substrate-runtime-joystream/pull/169

, and then we can open a new Issue for new ideas.
